test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/PromiseMock.scala

```diff
@@ -0,0 +1,225 @@
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation._
+import scala.scalajs.js.|
+
+import js.Thenable
+
+object PromiseMock {
+  import js.Dynamic.global
+
+  @noinline
+  def withMockedPromise[A](body: (() => Unit) => A): A = {
+    val oldPromise =
+      if (global.hasOwnProperty("Promise").asInstanceOf[Boolean]) Some(global.Promise)
+      else None
+
+    global.Promise = js.constructorOf[MockPromise[_]]
+    try {
+      body(MockPromise.processQueue _)
+    } finally {
+      oldPromise.fold {
+        global.asInstanceOf[js.Dictionary[Any]].delete("Promise")
+      } { old =>
+        global.Promise = old
+      }
+    }
+  }
+
+  @noinline
+  def withMockedPromiseIfExists[A](body: (Option[() => Unit]) => A): A = {
+    val oldPromise = global.Promise
+
+    if (js.isUndefined(oldPromise)) {
+      body(None)
+    } else {
+      global.Promise = js.constructorOf[MockPromise[_]]
+      try {
+        body(Some(MockPromise.processQueue _))
+      } finally {
+        global.Promise = oldPromise
+      }
+    }
+  }
+
+  private object MockPromise {
+    private val queue = js.Array[js.Function0[Any]]()
+
+    def enqueue(f: js.Function0[Any]): Unit =
+      queue.push(f)
+
+    def processQueue(): Unit = {
+      while (queue.nonEmpty)
+        queue.shift()()
+    }
+
+    private sealed abstract class State[+A]
+
+    private case object Pending extends State[Nothing]
+    private case class Fulfilled[+A](value: A) extends State[A]
+    private case class Rejected(reason: Any) extends State[Nothing]
+
+    private def isNotAnObject(x: Any): Boolean = x match {
+      case null | () | _:Double | _:Boolean | _:String => true
+      case _                                           => false
+    }
+
+    private def isCallable(x: Any): Boolean =
+      js.typeOf(x.asInstanceOf[js.Any]) == "function"
+
+    private def throwAny(e: Any): Nothing = {
+      throw (e match {
+        case th: Throwable => th
+        case _             => js.JavaScriptException(e)
+      })
+    }
+
+    private def tryCatchAny[A](tryBody: => A)(catchBody: Any => A): A = {
+      try {
+        tryBody
+      } catch {
+        case th: Throwable =>
+          catchBody(th match {
+            case js.JavaScriptException(e) => e
+            case _                         => th
+          })
+      }
+    }
+  }
+
+  @ScalaJSDefined
+  private class MockPromise[+A](
+      executor: js.Function2[js.Function1[A | Thenable[A], _], js.Function1[scala.Any, _], _])
+      extends js.Object with js.Thenable[A] {
+
+    import MockPromise._
+
+    private[this] var state: State[A] = Pending
+
+    private[this] var fulfillReactions = js.Array[js.Function1[A, Any]]()
+    private[this] var rejectReactions = js.Array[js.Function1[Any, Any]]()
+
+    init(executor)
+
+    private[this] def init(
+        executor: js.Function2[js.Function1[A | Thenable[A], _], js.Function1[scala.Any, _], _]) = {
+      tryCatchAny[Unit] {
+        executor(resolve _, reject _)
```

Comment:
http://www.ecma-international.org/ecma-262/6.0/#sec-promise-executor

I'll add the reference as a comment on `def init`, like I did for some other methods.